src/axom/primal/geometry/Quadrilateral.hpp

```diff
@@ -0,0 +1,264 @@
+// Copyright (c) 2017-2023, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#ifndef AXOM_PRIMAL_QUADRILATERAL_HPP_
+#define AXOM_PRIMAL_QUADRILATERAL_HPP_
+
+#include "axom/config.hpp"
+#include "axom/core.hpp"
+#include "axom/slic/interface/slic.hpp"
+#include "axom/fmt.hpp"
+
+#include "axom/primal/constants.hpp"
+#include "axom/primal/geometry/Point.hpp"
+#include "axom/primal/geometry/Triangle.hpp"
+
+#include <cmath>
+#include <ostream>
+
+namespace axom
+{
+namespace primal
+{
+// Forward declare the templated classes and operator functions
+template <typename T, int NDIMS>
+class Quadrilateral;
+
+/**
+ * \brief Overloaded output operator for quadrilaterals
+ */
+template <typename T, int NDIMS>
+std::ostream& operator<<(std::ostream& os, const Quadrilateral<T, NDIMS>& quad);
+
+/*!
+ * \accelerated
+ * \class Quadrilateral
+ *
+ * \brief Represents a quadrilateral geometric shape defined by four points.
+ *
+ * \accelerated
+ *
+ * \tparam T the coordinate type, e.g., double, float, etc.
+ * \tparam NDIMS the number of dimensions
+ *
+ * There are four vertices in the quadrilateral, labeled A through D as the
+ * constructor's arguments.  They are accessible using the square-brackets
+ * operator, with A being index 0, B index 1, C index 2, and D index 3.
+ *
+ * Here's a diagram showing a square with the labeled vertices.
+ *
+ * \verbatim
+ *
+ * D +---------+ C       +y
+ *   |         |         
+ *   |         |          ^
+ *   |         |          |
+ *   |         |          |
+ * A +---------+ B        -----> +x
+ *
+ * \endverbatim
```

Comment:
Nice ascii art!

Thanks for updating this to a right-handed coordinate system (#1184).